Sql query endpoint

npawar opened the pull request:
https://github.com/apache/incubator-pinot/issues/4962

Query the port using
```
curl -H "Content-Type: application/json" -X POST -d '{"sql":"select count(*) from testTable group by foo order by foo desc"}' http://localhost:8099/sql
```
Please not: the functionality will be limited right now by what's accepted by the parsing layer.
For instance, `TOP` will not parse for group by queries, but `LIMIT` - which will parse - hasn't been wired in yet.

mayankshriv commented on pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotClientRequest.java:
```diff
@@ -103,4 +100,60 @@ public String processQueryPost(String query) {
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  @Path("sql")
+  @ApiOperation(value = "Querying pinot in sql")
```
Nit: s/in/using?

jackjlli commented on pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotClientRequest.java:
```diff
@@ -103,4 +100,60 @@ public String processQueryPost(String query) {
+  @Path("sql")
+  @ApiOperation(value = "Querying pinot")
+  @ApiResponses(value = {@ApiResponse(code = 200, message = "Query response"), @ApiResponse(code = 500, message = "Internal Server Error")})
+  public String processSqlQueryPost(String query) {
```
Why not contain trace and debug_option in this endpoint?

npawar replied:
Everything is part of the payload in this one

jackjlli commented on pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotClientRequest.java:
```diff
@@ -103,4 +100,60 @@ public String processQueryPost(String query) {
+  @POST
+  @Produces(MediaType.APPLICATION_JSON)
+  @Path("sql")
+  @ApiOperation(value = "Querying pinot")
```
`"Querying pinot using sql"`?

npawar commented:
> Could you add a test as well?

Modified existing `testHardcodedSql` method in our integration tests, to start using the new endpoint.
It contains a very basic query set as of now. I've added TODO to complete that. It can be enriched as and when we start testing the new endpoint and certifying queries.

mayankshriv commented on pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ResultTableResultSet.java:
```diff
@@ -0,0 +1,106 @@
+class ResultTableResultSet extends AbstractResultSet {
+  private JsonNode _rowsArray;
+  private JsonNode _columnNamesArray;
+  private JsonNode _columnDataTypesArray;
```
Can we replace name and type with schema, and schema can be within metadata field. This way, we can add more metadata as needed.

npawar replied:
I feel this is okay as is, because:
The purpose of these ResultSets is to get the fields. For a user using the pinot client, once the ResultTable gets wrapped into ResultSet, the ResultSet will be used directly for accessing data. It will not really be converted to other forms further. So this is just internal representation.

For metadata, this is easy to change later. That will only be necessary when more APIs are added to the ResultSet. The current set of APIs are answered fine with this structure.

Lmk if you think it's okay.

siddharthteotia commented on pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransport.java:
```diff
@@ -67,9 +67,16 @@ public BrokerResponse executeQuery(String brokerAddress, String query)
+      json.put(queryFormat, request.getQuery());
+
+      final String url;
+      if (queryFormat.equalsIgnoreCase("sql")) {
```
I feel this is not very intuitive as to why the url path has /query for PQL query format and /sql for SQL query format. May be both modes can have query prefix? I don't think we can change the current url format as that would be backward compatible but for the SQL format, I think we should consider having /query/sql

npawar replied:
I had considered "/query/sql". But felt it was okay to go with just "sql", as we would eventually be getting rid of the "/query" endpoint.
I don't have a strong preference either way. I can change it if majority think this one is better. I got a signoff on the issue #4962 for "/sql"

npawar replied:
changed thi to /query/sql

siddharthteotia commented on pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ResultTableResultSet.java:
```diff
@@ -0,0 +1,106 @@
+ * ResultSet which contains the ResultTable from the broker response of a sql query.
+ */
+class ResultTableResultSet extends AbstractResultSet {
+  private JsonNode _rowsArray;
```
(nit): should be final

siddharthteotia commented:
LGTM
